RecoLocalCalo/HcalRecProducers/plugins/alpaka/Mahi.dev.cc

```diff
@@ -0,0 +1,1470 @@
+                                                                     diffVarItvlIdxZeroVec)
+                                         : 0;
+                  auto const value_t0m = idx >= 0 && static_cast<unsigned>(idx) < nsamples
+                                             ? compute_pulse_shape_value(t0m,
+                                                                         idx,
+                                                                         shift,
+                                                                         acc25nsVec,
+                                                                         diff25nsItvlVec,
+                                                                         accVarLenIdxMinusOneVec,
+                                                                         diffVarItvlIdxMinusOneVec,
+                                                                         accVarLenIdxZeroVec,
+                                                                         diffVarItvlIdxZeroVec)
+                                             : 0;
+                  auto const value_t0p = idx >= 0 && static_cast<unsigned>(idx) < nsamples
+                                             ? compute_pulse_shape_value(t0p,
+                                                                         idx,
+                                                                         shift,
+                                                                         acc25nsVec,
+                                                                         diff25nsItvlVec,
+                                                                         accVarLenIdxMinusOneVec,
+                                                                         diffVarItvlIdxMinusOneVec,
+                                                                         accVarLenIdxZeroVec,
+                                                                         diffVarItvlIdxZeroVec)
+                                             : 0;
+
+                  // store to global
+                  if (amplitude > 0.f) {
+                    pulseMatrix[ipulse * nsamples + sample] = value;
+                    pulseMatrixM[ipulse * nsamples + sample] = value_t0m;
+                    pulseMatrixP[ipulse * nsamples + sample] = value_t0p;
+                  } else {
+                    pulseMatrix[ipulse * nsamples + sample] = 0.f;
+                    pulseMatrixM[ipulse * nsamples + sample] = 0.f;
+                    pulseMatrixP[ipulse * nsamples + sample] = 0.f;
+                  }
+
+                }  // loop over sample
+              }    // loop over pulse
+            }      // loop over channels
+          }        // loop over group of channels
+        }
+      };  // Kernel_prep_pulseMatrices_sameNumberOfSamples
+
+      template <int NSAMPLES, int NPULSES>
+      ALPAKA_FN_ACC ALPAKA_FN_INLINE void update_covariance(
```

Comment:
Also in this function, `covarianceMatrix` is built in shared memory, but the different threads read from and write to it without any synchronisation ?